### PR TITLE
checkpointer: batch writes in RAM until block save

### PR DESCRIPTION
**Pull request by gregorycollins**

### CODE REVIEW

**Comment by gregorycollins** [comment]
Initial benchmark results from Emmanuel:

```
emmanuel@emmanuel-ThinkPad-P52:~/work/chainweb-node$ cabal new-run bench
Build profile: -w ghc-8.4.4 -O1
In order, the following will be built (use -v for more details):
 - chainweb-0.1.0.0 (bench:bench) (file src/Chainweb/Pact/Backend/Bench.hs changed)
Preprocessing library for chainweb-0.1.0.0..
Building library for chainweb-0.1.0.0..
[140 of 140] Compiling Chainweb.Pact.Backend.Bench ( src/Chainweb/Pact/Backend/Bench.hs, /home/emmanuel/work/chainweb-node/dist-newstyle/build/x86_64-linux/ghc-8.4.4/chainweb-0.1.0.0/build/Chainweb/Pact/Backend/Bench.o )
Preprocessing benchmark 'bench' for chainweb-0.1.0.0..
Building benchmark 'bench' for chainweb-0.1.0.0..
Linking /home/emmanuel/work/chainweb-node/dist-newstyle/build/x86_64-linux/ghc-8.4.4/chainweb-0.1.0.0/build/bench/bench ...
benchmarking pact-backend/pact-sqlite/transactionCount=1/usertable
time                 33.35 μs   (32.83 μs .. 34.12 μs)
                     0.997 R²   (0.994 R² .. 0.999 R²)
mean                 34.34 μs   (33.21 μs .. 38.35 μs)
std dev              6.932 μs   (1.300 μs .. 14.59 μs)
variance introduced by outliers: 96% (severely inflated)

benchmarking pact-backend/pact-sqlite/transactionCount=5/usertable
time                 167.8 μs   (166.3 μs .. 170.0 μs)
                     0.998 R²   (0.995 R² .. 1.000 R²)
mean                 167.7 μs   (166.3 μs .. 169.8 μs)
std dev              6.973 μs   (4.296 μs .. 11.14 μs)
variance introduced by outliers: 40% (moderately inflated)

benchmarking pact-backend/pact-sqlite/transactionCount=10/usertable
time                 327.8 μs   (323.6 μs .. 333.1 μs)
                     0.994 R²   (0.983 R² .. 0.999 R²)
mean                 328.5 μs   (322.4 μs .. 343.3 μs)
std dev              31.74 μs   (6.997 μs .. 59.92 μs)
variance introduced by outliers: 76% (severely inflated)

benchmarking pact-backend/pact-sqlite/transactionCount=20/usertable
time                 656.1 μs   (652.6 μs .. 661.6 μs)
                     0.998 R²   (0.993 R² .. 1.000 R²)
mean                 657.9 μs   (651.5 μs .. 680.1 μs)
std dev              30.24 μs   (9.560 μs .. 58.97 μs)
variance introduced by outliers: 38% (moderately inflated)

benchmarking pact-backend/pact-sqlite/transactionCount=50/usertable
time                 1.905 ms   (1.728 ms .. 2.077 ms)
                     0.969 R²   (0.948 R² .. 0.995 R²)
mean                 1.722 ms   (1.688 ms .. 1.789 ms)
std dev              150.6 μs   (77.64 μs .. 282.7 μs)
variance introduced by outliers: 64% (severely inflated)

benchmarking pact-backend/pact-sqlite/transactionCount=100/usertable
time                 3.454 ms   (3.368 ms .. 3.617 ms)
                     0.995 R²   (0.989 R² .. 1.000 R²)
mean                 3.367 ms   (3.353 ms .. 3.420 ms)
std dev              88.85 μs   (25.52 μs .. 181.6 μs)
variance introduced by outliers: 12% (moderately inflated)

benchmarking pact-backend/pact-sqlite/transactionCount=200/usertable
time                 7.130 ms   (7.104 ms .. 7.154 ms)
                     1.000 R²   (0.999 R² .. 1.000 R²)
mean                 7.089 ms   (7.069 ms .. 7.119 ms)
std dev              75.96 μs   (48.10 μs .. 127.6 μs)

benchmarking pact-backend/pact-sqlite/transactionCount=500/usertable
time                 20.84 ms   (20.27 ms .. 21.88 ms)
                     0.991 R²   (0.976 R² .. 1.000 R²)
mean                 20.51 ms   (20.26 ms .. 21.18 ms)
std dev              942.7 μs   (421.1 μs .. 1.737 ms)
variance introduced by outliers: 18% (moderately inflated)

benchmarking pact-backend/pact-sqlite/transactionCount=1000/usertable
time                 47.86 ms   (47.67 ms .. 48.28 ms)
                     1.000 R²   (0.999 R² .. 1.000 R²)
mean                 47.68 ms   (47.52 ms .. 48.09 ms)
std dev              468.7 μs   (172.1 μs .. 795.2 μs)

benchmarking pact-backend/batchedCheckpointer/transactionCount=1/usertable
time                 194.6 μs   (193.9 μs .. 195.8 μs)
                     1.000 R²   (0.999 R² .. 1.000 R²)
mean                 194.6 μs   (194.2 μs .. 195.4 μs)
std dev              1.988 μs   (1.041 μs .. 3.346 μs)

benchmarking pact-backend/batchedCheckpointer/transactionCount=5/usertable
time                 243.6 μs   (243.3 μs .. 243.9 μs)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 245.5 μs   (243.8 μs .. 250.9 μs)
std dev              8.779 μs   (4.545 μs .. 16.93 μs)
variance introduced by outliers: 32% (moderately inflated)

benchmarking pact-backend/batchedCheckpointer/transactionCount=10/usertable
time                 295.7 μs   (292.9 μs .. 301.5 μs)
                     0.999 R²   (0.997 R² .. 1.000 R²)
mean                 294.2 μs   (292.7 μs .. 297.9 μs)
std dev              7.256 μs   (1.758 μs .. 12.36 μs)
variance introduced by outliers: 18% (moderately inflated)

benchmarking pact-backend/batchedCheckpointer/transactionCount=20/usertable
time                 377.1 μs   (374.8 μs .. 381.2 μs)
                     0.999 R²   (0.999 R² .. 1.000 R²)
mean                 376.2 μs   (374.7 μs .. 379.1 μs)
std dev              6.325 μs   (2.784 μs .. 10.74 μs)

benchmarking pact-backend/batchedCheckpointer/transactionCount=50/usertable
time                 632.1 μs   (625.9 μs .. 642.6 μs)
                     0.999 R²   (0.997 R² .. 1.000 R²)
mean                 628.5 μs   (625.7 μs .. 633.3 μs)
std dev              12.64 μs   (6.201 μs .. 20.81 μs)
variance introduced by outliers: 11% (moderately inflated)

benchmarking pact-backend/batchedCheckpointer/transactionCount=100/usertable
time                 1.117 ms   (1.083 ms .. 1.159 ms)
                     0.993 R²   (0.985 R² .. 1.000 R²)
mean                 1.093 ms   (1.081 ms .. 1.119 ms)
std dev              55.21 μs   (17.14 μs .. 89.80 μs)
variance introduced by outliers: 39% (moderately inflated)

benchmarking pact-backend/batchedCheckpointer/transactionCount=200/usertable
time                 2.028 ms   (1.983 ms .. 2.101 ms)
                     0.996 R²   (0.990 R² .. 1.000 R²)
mean                 1.948 ms   (1.928 ms .. 1.972 ms)
std dev              71.62 μs   (42.88 μs .. 122.9 μs)
variance introduced by outliers: 23% (moderately inflated)

benchmarking pact-backend/batchedCheckpointer/transactionCount=500/usertable
time                 4.831 ms   (4.642 ms .. 5.063 ms)
                     0.993 R²   (0.988 R² .. 0.999 R²)
mean                 4.631 ms   (4.594 ms .. 4.720 ms)
std dev              159.0 μs   (72.55 μs .. 267.1 μs)
variance introduced by outliers: 16% (moderately inflated)

benchmarking pact-backend/batchedCheckpointer/transactionCount=1000/usertable
time                 9.587 ms   (9.192 ms .. 9.981 ms)
                     0.995 R²   (0.990 R² .. 1.000 R²)
mean                 9.209 ms   (9.142 ms .. 9.341 ms)
std dev              259.1 μs   (106.9 μs .. 444.8 μs)
```

File: src/Chainweb/Pact/Backend/Bench.hs [hunk 1]
```diff
@@ -56,9 +65,78 @@ cpBench = C.envWithCleanup setup teardown $ \ ~(NoopNFData (e,_)) -> C.bgroup "c
...
+        k = "k"
+
+        hash01 = BlockHash $ unsafeMerkleLogHash "0000000000000000000000000000001a"
+        hash02 = BlockHash $ unsafeMerkleLogHash "0000000000000000000000000000002a"
+
+        writeRow (PactDbEnv pdb e) writeType ut i =
+            _writeRow pdb writeType ut k (ObjectMap $ M.fromList [(f,(PLiteral (LInteger i)))]) e
+
+        go CheckpointEnv{..} (NoopNFData ut) = do
+            result <- restore _cpeCheckpointer (Just (BlockHeight 1, hash01)) >>= \case
```
Comment (slpopejoy) [inline src/Chainweb/Pact/Backend/Bench.hs]: Interesting, so this benchmark does a rewind each time. We should make a variant that has an MVar with an increasing height/hash to test "non-rewind performance" too.

File: src/Chainweb/Pact/Backend/Types.hs [hunk 2]
```diff
@@ -141,6 +155,33 @@ instance FromJSON PactDbConfig
 
 makeLenses ''PactDbConfig
 
+data SQLiteRowDelta = SQLiteRowDelta
```
Comment (slpopejoy) [inline src/Chainweb/Pact/Backend/Types.hs]: Haddock all of these: SQLite[RowDelta,PendingWrites,PendingData] in order to  describe their semantics

File: src/Chainweb/Pact/Backend/Bench.hs [hunk 3]
```diff
@@ -33,20 +33,120 @@ import Chainweb.BlockHash
...
+        (initbytestring, hash01) = (bytestring, BlockHash $ unsafeMerkleLogHash bytestring)
+            where
+              bytestring =
+                "0000000000000000000000000000001a"
+
+        nextHash bytestring =
+          (bytestring, BlockHash $ unsafeMerkleLogHash $ B.pack $ B.zipWith (+) bytestring inc)
+          where
+            inc =
+              "\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\SOH\NUL"
```
Comment (gregorycollins) [inline src/Chainweb/Pact/Backend/Bench.hs]: yick -- make this `B.replicate k 0 <> "\SOH\NUL"`

File: src/Chainweb/Pact/Backend/Types.hs (hunk 2 above)
Comment (gregorycollins) [reply src/Chainweb/Pact/Backend/Types.hs]: done

File: src/Chainweb/Pact/Backend/Bench.hs (hunk 1 above)
Comment (gregorycollins) [reply src/Chainweb/Pact/Backend/Bench.hs]: Done

File: src/Chainweb/Pact/Backend/Bench.hs (hunk 3 above)
Comment (gregorycollins) [reply src/Chainweb/Pact/Backend/Bench.hs]: ... but we'll do that after merge

**Pull request merged**